core/state_transition.go

```diff
@@ -494,6 +529,70 @@ func (st *stateTransition) execute() (*ExecutionResult, error) {
 	}, nil
 }
 
+// validateAuthorization validates an EIP-7702 authorization against the state.
+func (st *stateTransition) validateAuthorization(auth *types.Authorization) (authority common.Address, err error) {
+	// Verify chain ID is 0 or equal to current chain ID.
+	if auth.ChainID != 0 && st.evm.ChainConfig().ChainID.Uint64() != auth.ChainID {
+		return authority, ErrAuthorizationWrongChainID
+	}
+	// Limit nonce to 2^64-1 per EIP-2681.
+	if auth.Nonce+1 < auth.Nonce {
+		return authority, ErrAuthorizationNonceOverflow
+	}
+	// Validate signature values and recover authority.
+	authority, err = auth.Authority()
+	if err != nil {
+		return authority, fmt.Errorf("%w: %v", ErrAuthorizationInvalidSignature, err)
+	}
+	// Check the authority account
+	//  1) doesn't have code or has exisiting delegation
+	//  2) matches the auth's nonce
+	//
+	// Note it is added to the access list even if the authorization is invalid.
+	st.state.AddAddressToAccessList(authority)
+	code := st.state.GetCode(authority)
+	if _, ok := types.ParseDelegation(code); len(code) != 0 && !ok {
+		return authority, ErrAuthorizationDestinationHasCode
+	}
+	if have := st.state.GetNonce(authority); have != auth.Nonce {
+		return authority, ErrAuthorizationNonceMismatch
+	}
+	return authority, nil
+}
+
+// applyAuthorization applies an EIP-7702 code delegation to the state.
+func (st *stateTransition) applyAuthorization(msg *Message, auth *types.Authorization) error {
+	authority, err := st.validateAuthorization(auth)
+	if err != nil {
+		return err
+	}
+
+	// If the account already exists in state, refund the new account cost
+	// charged in the intrinsic calculation.
+	if st.state.Exist(authority) {
+		st.state.AddRefund(params.CallNewAccountGas - params.TxAuthTupleGas)
+	}
+
+	// Update nonce and account code.
+	st.state.SetNonce(authority, auth.Nonce+1)
+	delegation := types.AddressToDelegation(auth.Address)
+	if auth.Address == (common.Address{}) {
+		// If the delegation is for the zero address, completely clear all
+		// delegations from the account.
+		delegation = []byte{}
+	}
+	st.state.SetCode(authority, delegation)
+
+	// Usually the transaction destination and delegation target are added to
+	// the access list in statedb.Prepare(..), however if the delegation is in
+	// the same transaction we need add here as Prepare already happened.
+	if *msg.To == authority {
+		st.state.AddAddressToAccessList(auth.Address)
+	}
```

Comment:
This is a bit unintuitive. 
So we check equality between `msg.To` and `authority`. But we use a third form, `auth.Address`, when we add it to the address list. Why not just `st.state.AddAddressToAccessList(authority)`? That _is_ the same, right? 

But I still don't get it. the `msg.To`, the destination address, is added during `Prepare`. Why do we have to special-case add it once again here?

Reply:
I have expanded the comment

Reply:
`auth.Address` is actually different than `authority`. It is the address that `authority` *points` to, so it is address of the code they have authorized to execute in their account. 

We need this special case here because our prepare function is called before we can handle the authorizations, which means we've already done the convenience warming for the target of `msg.To` and its destination. So we need to do it again here in case `msg.To` set a delegation in the same tx.